### DISCUSSION TITLE
Update part of a Custom variable

### DISCUSSION THREAD

**Discussion by psheehymn** [gapitio/gapit-htmlgraphics-panel#155] [Q&A]
I have several Custom variables which take advantage of the "text : value" format of that construct. I know I can retrieve the value of one half or the other by coding the JSON for the Custom Properties like so:

   "QPath01text": "${QPath01:text}",
   "QPath01value": "${QPath01:value}",

However, my efforts to _set_ one half or the other (or set both sides at the same time) in the JavaScript have thus far not worked. Trying to set either half (using the Custom Properties aliases above) results in no change to the variable. If I add a third alias to the Custom Properties for the unqualified variable, as in:

   "QPath01var": "${QPath01}",

and attempt a change, the provided value is inserted on both halves of the variable. This includes if I specify the value as 

"newText : newValue"

The resulting variable value is "newText : newValue : newText : newValue"

If there any syntax that I can use to successfully set one side or the other (or both sides at the same time to separate values) or is this not possible and I will need to use separate Constant variables?

**Comment by ZuperZee** [comment]
Can you provide an example dashboard/panel of what you've tried or how the Grafana variables are set up?
I might be misreading, but are you trying to update the Grafana variable or the Custom Properties?
If you are trying to update the Grafana variable, are you trying to update the value and the text?

**Comment by psheehymn** [comment]
I am trying to update the Grafana variable. I want to update both the text and value, but to different values.

For example, I want to update ${QPath01:text} to “Dashboard” and ${QPath01:value} to “/dashboard”. I can code it such that I update them independently (preferred) or I can update them at the same time (i.e. with the same call to htmlGraphics.locationService.partial)  if need be.

I can try to put together a simple test panel if the above clarification is insufficient.

Thanks,

Pat

Patrick M. Sheehy | Portal Monitoring Program Management, Optum Digital
M (507) 251-9899
***@***.******@***.***> | www.myuhc.com<http://www.myuhc.com/>

Helping People Live Healthier Lives®
■ Integrity ■ Compassion ■ Relationships ■ Innovation ■ Performance

**Reply by ZuperZee** [reply]
You can update the Grafana variable value as shown in this example: https://gapit-htmlgraphics-panel.gapit.io/docs/examples/update-grafana-variable-example/
But you can't update the Grafana variable name or label without using the Grafana API to update the entire dashboard and change the template variable there. It would require a fetch/post command to the Grafana API, which requires the user to have edit permissions on the dashboard.

**Comment by psheehymn** [comment]
The example shows updating a custom variable that contains only a list of values. My custom variable has a single “text : value” pair. I am not changing the variable name or label, only the content of the variable, but I want to be able to change each side of the content pair.

Patrick M. Sheehy | Portal Monitoring Program Management, Optum Digital
M (507) 251-9899
***@***.******@***.***> | www.myuhc.com<http://www.myuhc.com/>

Helping People Live Healthier Lives®
■ Integrity ■ Compassion ■ Relationships ■ Innovation ■ Performance

**Reply by ZuperZee** [reply]
Can you share a screenshot of it, like shown below?
![image](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/ed2ffa7c-f0c7-4b55-8c64-db6c1e58f460)

**Comment by psheehymn** [comment]
Here's the variable panel:

![image](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/93273674/a4ad29eb-9057-4aea-93aa-bfde2620f2a9)

I've also attached the JSON for a test dashboard that tries the four ways I experimented with.
[HTMLG Custom Variable Test-1691592094898.json.txt](https://github.com/gapitio/gapit-htmlgraphics-panel/files/12303937/HTMLG.Custom.Variable.Test-1691592094898.json.txt)

**Reply by ZuperZee** [reply]
I couldn't get a working solution for it. I created a community post https://community.grafana.com/t/set-a-key-value-grafana-variable/100832
It looks like a Grafana issue, but I'll wait a bit to see if someone knows.

Seems like Grafana doesn't allow the key : value syntax in locationService.
![image](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/2145ebb4-a55e-4f53-adbd-5f9a67990824)
![update-key-value-grafana-variable](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/94acd911-4398-42ad-a001-595c1e444ca5)
[dashboard.json.txt](https://github.com/gapitio/gapit-htmlgraphics-panel/files/12306017/dashboard.json.txt)

BTW. You can reply to your own "answer" instead of creating new ones.
![image](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/dd4616f3-6b4f-46f4-be83-dd308f0843fe)

**Reply by psheehymn** [reply]
Thanks! I had overlooked this box previously, as it was initially collapsed.

**Reply by psheehymn** [reply]
Thank you the update and for following up in the community forum. I'll use separate variables in the interim until such time as there is a solution in the locationService.

**Reply by ZuperZee** [reply]
Someone responded: https://community.grafana.com/t/set-a-key-value-grafana-variable/100832/2. Could you respond to the question?

The question in the post:
I guess my question is what is the end goal of this effort and how will you use these dynamic key values?

**Reply by psheehymn** [reply]
Answered in the community forum item, but happy to duplicate here for completeness if desired.

**Reply by ZuperZee** [reply]
Nah. No need.

**Reply by psheehymn** [reply]
@ZuperZee, could you please review the chain in [https://community.grafana.com/t/set-a-key-value-grafana-variable/100832](https://community.grafana.com/t/set-a-key-value-grafana-variable/100832) to see if you have any additional insights to offer? 

The various respondents seem to suggest that what I am trying to do should work. Since the function call to `locationService.partial` is prefaced with `htmlGraphics`, I assume the function is being imported into your **htmlGraphics** class from **@grafana/runtime**? Are you extending or otherwise changing the function in any way, or is it a straight pass-through?  I will confess to having very limited background in Grafana plugin construction, so if this is a stupid question, please accept my apology. I'm "grasping at straws" so to speak trying to figure this problem out. Thank you in advance for any additional assistance you can offer.
